File: insight-brain-data/src/main/java/com/sonatype/insight/brain/model/configuration/saml/SamlConfiguration.java
/*
 * Copyright (c) 2011-present Sonatype, Inc. All rights reserved.
 * Includes the third-party code listed at http://links.sonatype.com/products/clm/attributions.
 * "Sonatype" is a trademark of Sonatype, Inc.
 */
package com.sonatype.insight.brain.model.configuration.saml;

import java.security.KeyPair;
import java.security.PrivateKey;
import java.security.cert.Certificate;

import com.fasterxml.jackson.annotation.JsonIgnore;

/**
 * @since 1.72
 */
public class SamlConfiguration
{
  public static final int IDENTITY_PROVIDER_NAME_MAXIMUM_LENGTH = 200;

  private String id;

  private String identityProviderName = "identity provider";

  private String identityProviderMetadataXml;

  private String entityId;

  private String firstNameAttributeName = "firstName";

  private String lastNameAttributeName = "lastName";

  private String emailAttributeName = "email";

  private String usernameAttributeName = "username";

  private String groupsAttributeName = "groups";

  private Boolean validateResponseSignature;

  private Boolean validateAssertionSignature;

  @JsonIgnore
  private Certificate certificate;

  @JsonIgnore
  private PrivateKey decryptionKey;

  @JsonIgnore
  private KeyPair signingKeyPair;

  public String getId() {
    return id;
  }

  public void setId(String id) {
    this.id = id;
  }

  public String getIdentityProviderName() {
    return identityProviderName;
  }

  public void setIdentityProviderName(String identityProviderName) {
    this.identityProviderName = identityProviderName;
  }

  public String getFirstNameAttributeName() {
    return firstNameAttributeName;
  }

  public void setFirstNameAttributeName(String firstNameAttributeName) {
    this.firstNameAttributeName = firstNameAttributeName;
  }

  public String getLastNameAttributeName() {
    return lastNameAttributeName;
  }

  public void setLastNameAttributeName(String lastNameAttributeName) {
    this.lastNameAttributeName = lastNameAttributeName;
  }

  public String getEmailAttributeName() {
    return emailAttributeName;
  }

  public void setEmailAttributeName(String emailAttributeName) {
    this.emailAttributeName = emailAttributeName;
  }

  public String getUsernameAttributeName() {
    return usernameAttributeName;
  }

  public void setUsernameAttributeName(String usernameAttributeName) {
    this.usernameAttributeName = usernameAttributeName;
  }

  public String getGroupsAttributeName() {
    return groupsAttributeName;
  }

  public void setGroupsAttributeName(String groupsAttributeName) {
    this.groupsAttributeName = groupsAttributeName;
  }

  public String getIdentityProviderMetadataXml() {
    return identityProviderMetadataXml;
  }

  public void setIdentityProviderMetadataXml(String identityProviderMetadataXml) {
    this.identityProviderMetadataXml = identityProviderMetadataXml;
  }

  public String getEntityId() {
    return entityId;
  }

  public void setEntityId(String entityId) {
    this.entityId = entityId;
  }

  public Boolean getValidateResponseSignature() {
    return validateResponseSignature;
  }

  public void setValidateResponseSignature(Boolean validateResponseSignature) {
    this.validateResponseSignature = validateResponseSignature;
  }

  public Boolean getValidateAssertionSignature() {
    return validateAssertionSignature;
  }

  public void setValidateAssertionSignature(Boolean validateAssertionSignature) {
    this.validateAssertionSignature = validateAssertionSignature;
  }

  public Certificate getCertificate() {
    return certificate;
  }

  public void setCertificate(Certificate certificate) {
    this.certificate = certificate;
  }

  public PrivateKey getDecryptionKey() {
    return decryptionKey;
  }

  public void setDecryptionKey(PrivateKey decryptionKey) {
    this.decryptionKey = decryptionKey;
  }

  public KeyPair getSigningKeyPair() {
    return signingKeyPair;
  }

  public void setSigningKeyPair(KeyPair signingKeyPair) {
    this.signingKeyPair = signingKeyPair;
  }
}

File: insight-brain-frontend/src/main/frontend/configuration/saml/samlConfiguration.js
/*
 * Copyright (c) 2011-present Sonatype, Inc. All rights reserved.
 * Includes the third-party code listed at http://links.sonatype.com/products/clm/attributions.
 * "Sonatype" is a trademark of Sonatype, Inc.
 */
import template from './samlConfiguration.html';
import { omit } from 'ramda';

export default {
  template: template,
  bindings: {
    isAuthorized: '<',
  },
  controllerAs: 'vm',
  controller: SamlConfigurationController,
};

function SamlConfigurationController($scope, BaseUrl, $http, CLMContextLocations, Messages, Dialog, $window) {
  const vm = this;
  const defaultSaml = {
    identityProviderName: 'identity provider',
    identityProviderMetadataXml: undefined,
    entityId: BaseUrl.get() + '/api/v2/config/saml/metadata',
    usernameAttributeName: 'username',
    firstNameAttributeName: 'firstName',
    lastNameAttributeName: 'lastName',
    emailAttributeName: 'email',
    groupsAttributeName: 'groups',
    validateResponseSignature: null,
    validateAssertionSignature: null,
  };
  const validateSignatureOptions = [
    { name: 'Default', value: null },
    { name: 'True', value: true },
    { name: 'False', value: false },
  ];

  let originalSaml = undefined;

  Object.assign(vm, {
    loadError: undefined,
    saveOrDeleteError: undefined,
    saml: undefined,
    defaultSaml: angular.copy(defaultSaml),
    isUpdating: undefined,
    samlConfigurationMask: undefined,
    validateSignatureOptions: angular.copy(validateSignatureOptions),
    load() {
      resetErrors();
      $http.get(CLMContextLocations.getSamlConfigurationUrl()).then(
        function (response) {
          setSaml(response.data, true);
        },
        function (error) {
          if (error.status === 404) {
            setSaml(defaultSaml, false);
            return;
          }
          onError('loadError', error);
        }
      );
    },
    readIdentityProviderMetadataXml(file) {
      if (file !== undefined) {
        getFileReader().readAsText(file);
      }
    },
    isChanged() {
      return !angular.equals(vm.saml, originalSaml);
    },
    save() {
      resetErrors();
      let formData = new FormData();
      formData.append('identityProviderXml', vm.saml.identityProviderMetadataXml);
      let payload = omit(['identityProviderMetadataXml'], vm.saml);
      formData.append('samlConfiguration', JSON.stringify(payload));
      vm.samlConfigurationMask
        .wrap(
          $http
            .put(CLMContextLocations.getSamlConfigurationUrl(), formData, {
              // Angular's default transformRequest will try to serialize our formData, so we override it with the identity
              // function to leave formData intact
              transformRequest: angular.identity,
              // Angular's default Content-Type header for POST/PUT is application/json, by setting it to undefined the
              // browser sets it to multipart/form-data and fills in the correct boundary (which wouldn't happen if we set it
              // manually to multipart/form-data)
              headers: { 'Content-Type': undefined },
            })
            .then(vm.load)
        )
        .catch((error) => onError('saveOrDeleteError', error));
    },
    cancel() {
      resetErrors();
      vm.saml = angular.copy(originalSaml);
    },
    delete() {
      Dialog.open({
        title: 'Delete Configuration',
        body: 'Are you sure you want to delete this SAML configuration?',
        id: 'delete-saml-confirmation',
        buttons: [
          {
            name: 'Delete',
            type: 'primary',
            click: deleteConfiguration,
          },
          {
            name: 'Cancel',
            type: 'cancel',
          },
        ],
      });
    },
    defaultsToTooltipText(defaultValue) {
      return 'If empty will default to "' + defaultValue + '"';
    },
    resetToDefaultValueIfEmpty(name) {
      if (!vm.saml[name]) {
        vm.saml[name] = defaultSaml[name];
      }
    },
    // IE workaround
    downloadMetadataForIE() {
      if (vm.isUpdating && $window.navigator.msSaveBlob) {
        $http
          .get(CLMContextLocations.getSamlConfigurationUrl() + '/metadata')
          .then((response) => $window.navigator.msSaveBlob(new Blob([response.data]), 'metadata.xml'));
      }
    },
    shouldEnableDownloadMetadataLink() {
      return vm.isUpdating && !$window.navigator.msSaveBlob;
    },
  });

  function getFileReader() {
    let fileReader = new FileReader();
    fileReader.addEventListener('load', function () {
      $scope.$apply(function () {
        vm.saml.identityProviderMetadataXml = fileReader.result;
      });
    });
    return fileReader;
  }

  function deleteConfiguration() {
    resetErrors();
    vm.samlConfigurationMask
      .wrap($http.delete(CLMContextLocations.getSamlConfigurationUrl()).then(vm.load))
      .catch((error) => onError('saveOrDeleteError', error));
  }

  function setSaml(saml, exists) {
    vm.saml = angular.copy(saml);
    originalSaml = angular.copy(vm.saml);
    vm.isUpdating = exists;
  }

  function onError(target, error) {
    vm[target] = Messages.getHttpErrorMessage(error);
  }

  function resetErrors() {
    vm.loadError = undefined;
    vm.saveOrDeleteError = undefined;
  }

  vm.load();
}

SamlConfigurationController.$inject = [
  '$scope',
  'BaseUrl',
  '$http',
  'CLMContextLocations',
  'Messages',
  'Dialog',
  '$window',
];
